From c25f37d6f18a00bedc897f2cbe15085c7a6b8d34 Mon Sep 17 00:00:00 2001
From: Roger Burtonpatel <roger.burtopatel@tufts.edu>
Date: Wed, 8 Mar 2023 15:52:21 -0500
Subject: [PATCH] added unicode %U to VM printer

---
 src/svm/print.c | 37 +++++++++++++++++++++++++++++++++++++
 src/svm/vmrun.c |  6 ++++++
 2 files changed, 43 insertions(+)

diff --git a/src/svm/print.c b/src/svm/print.c
index 1a84223..bb101f4 100644
--- a/src/svm/print.c
+++ b/src/svm/print.c
@@ -11,6 +11,7 @@
 #include "print.h"
 #include "value.h"
 #include "vmstring.h"
+#include "vmerror.h"
 
 
 void bprint(Printbuf output, const char *fmt, ...) {
@@ -67,6 +68,41 @@ void vbprint(Printbuf output, const char *fmt, va_list_box *box) {
         }
     }
 }
+
+
+void printunicode(Printbuf output, va_list_box *box)
+{
+    unsigned code_point = va_arg(box->ap, unsigned);
+
+    if ((code_point & 0x1fffff) != code_point)
+        runerror(NULL, "%d does not represent a Unicode code point", 
+                        (int)code_point);
+    if (code_point > 0xffff) {     // 21 bits
+        bufput(output, 0xf0 |  (code_point >> 18));
+        bufput(output, 0x80 | ((code_point >> 12) & 0x3f));
+        bufput(output, 0x80 | ((code_point >>  6) & 0x3f));
+        bufput(output, 0x80 | ((code_point      ) & 0x3f));
+    } else if (code_point > 0x7ff) { // 16 bits
+        bufput(output, 0xe0 | (code_point >> 12));
+        bufput(output, 0x80 | ((code_point >> 6) & 0x3f));
+        bufput(output, 0x80 | ((code_point     ) & 0x3f));
+    } else if (code_point > 0x7f) { // 12 bits
+        bufput(output, 0xc0 | (code_point >> 6));
+        bufput(output, 0x80 | (code_point & 0x3f));
+    } else {                        // 7 bits
+        bufput(output, code_point);
+    }
+}
+
+void fprint_utf8(FILE *output, unsigned code_point)
+{
+    fprint(output, "%U", code_point);
+}
+void print_utf8 (unsigned u)
+{
+    fprint_utf8(stdout, u);
+}
+
 void installprinter(unsigned char c, Printer *take_and_print) {
     printertab[c] = take_and_print;
 }
@@ -138,6 +174,7 @@ void printchar(Printbuf output, va_list_box *box) {
 void installprinters(void) {
     installprinter('c', printchar);
     installprinter('d', printdecimal);
+    installprinter('U', printunicode);
     installprinter(',', printcommadecimal);
     installprinter(';', print64commadecimal);
     installprinter('n', printname);
diff --git a/src/svm/vmrun.c b/src/svm/vmrun.c
index fb746ec..8f5817b 100644
--- a/src/svm/vmrun.c
+++ b/src/svm/vmrun.c
@@ -33,6 +33,9 @@
 #define CANDUMP 1
 
 void vmrun(VMState vm, struct VMFunction *fun) {
+    print("%x\n", 0xf09d9aaf);
+    print("%x\n", 0xf09d9aaf & 0x1fffff);
+    print("%U\n", (unsigned)0x1d9aaf);
     
     if (fun->size < 1) {
         return;
@@ -72,6 +75,9 @@ void vmrun(VMState vm, struct VMFunction *fun) {
             case Println:
                 print("%v\n", registers[uX(curr_instr)]);
                 break;
+            case Printu:
+                print("%U", (unsigned)AS_NUMBER(vm, registers[uX(curr_instr)]));
+                break;                
             case Check: {
                 v = literal_value(vm, uYZ(curr_instr));
                 check(vm, AS_CSTRING(vm, v), registers[uX(curr_instr)]);
-- 
2.31.1

